server/src/leaderboard.ts

```diff
@@ -0,0 +1,328 @@
+import { Chain } from '../shovel-config'
+import db from './db'
+
+type LeaderboardLine = {
+  problemId: string
+  owner: string
+}
+
+type SpeedScore = LeaderboardLine & { ticks: number }
+
+type SolvedScore = LeaderboardLine & { solved: number }
+
+type StreakScore = LeaderboardLine & { streak: number }
+
+type DailyLeaderboard = {
+  oneBody: SpeedScore[]
+  twoBody: SpeedScore[]
+  threeBody: SpeedScore[]
+  cumulative: SpeedScore[]
+}
+
+type Leaderboard = {
+  daily: Record<number, DailyLeaderboard>
+  allTime: {
+    mostSolved: SolvedScore[]
+    currentStreak: StreakScore[]
+    fastest: SpeedScore[]
+  }
+}
+
+export const leaderboard: Leaderboard = {
+  daily: {},
+  allTime: {
+    mostSolved: [],
+    currentStreak: [],
+    fastest: []
+  }
+}
+
+function beginningOfTodayInUTCSeconds(): number {
+  const now = new Date()
+  const startOfTodayUTC = new Date(
+    Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate())
+  )
+  return Math.floor(startOfTodayUTC.getTime() / 1000)
+}
+
+async function calculateDailyLeaderboard(day: number, chain: Chain) {
+  const result = await db.query(`
+  WITH fastest_times AS (
+    SELECT 
+        problem_id,
+        level,
+        MIN(ticks_in_this_match) AS fastest_time
+    FROM 
+        solver_solved
+    WHERE 
+        level IN (1, 2, 3)
+        AND day = ${day}
+        AND src_name = '${chain}'
+    GROUP BY 
+        problem_id, level
+  ),
+  cumulative_times AS (
+      SELECT 
+          problem_id,
+          SUM(ticks_in_this_match) AS total_time
+      FROM 
+          solver_solved
+      WHERE 
+          day = ${day}
+          AND src_name = '${chain}'
+      GROUP BY 
+          problem_id
+      HAVING 
+          COUNT(level) = 3
```

Comment:
have been switching around max body count in tests, latest playtest we tried with 5 bodies. I wonder if it's worth making this a parameter stored along with the chain info so when we deploy a new one we update there? Or at least might be good to mark in the code wherever the max number of bodies is relevant to keep track of where to change when need be?

Reply:
Cool, I'll add a var to keep track of level count. Would be cool to have that on the chain too. Could be a `finalLevel` boolean on the Solved event.